File: src/2017/advent5.py
#!/usr/bin/python

import sys
import io

def jump1(arr):
   steps = 0
   current = 0

   while current < len(arr) and current >= 0:
      offset = arr[current]

      arr[current] += 1
      current = current + offset
      steps += 1

   return steps

def jump2(arr):
   steps = 0
   current = 0

   while current < len(arr) and current >= 0:
      offset = arr[current]
      arr[current] += 1 - 2*int(offset>=3)
      current += offset
      steps += 1

   return steps

def parse(l):
   return [int(s) for s in l.split()]

if __name__ == '__main__':
   if len(sys.argv) != 2:
      print 'usage: %s <file name>' % sys.argv[0]
      sys.exit(1)
   fname = sys.argv[1]
   with open(fname, 'r') as f:
      lines = f.read()

      l = parse(lines)
      print 'result mode 1 = %d' % jump1(l)

      l = parse(lines)
      print 'result mode 2 = %d' % jump2(l)

File: src/2017/advent5.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

int jump(std::vector<int>& insts)
{
   long steps = 0;
   long current = 0;

   const int N = insts.size();

   while(current < N & current >= 0)
   {
      long offset = insts[current];
      insts[current] += 1 - 2*(offset>=3);
      current += offset;
      ++steps;
   }
   return steps;
}

int main(int argc, const char** argv)
{
   const std::string fname = argv[1];
   std::ifstream infile(fname);
   std::vector<int> instructions;
   int instruction;

   while(infile >> instruction)
   {
      instructions.push_back(instruction);
   }

   return jump(instructions);
}
